File: nowcoderjs/class2/heapSort.js
// 将一个数组变成大根堆
function toBigHeap(arr) {
    let size = 0;
    let len = arr.length;
    for(let i =0; i< len;i++) {
        let index = i;
        while(arr[index] > arr[Math.round((index-1)/2)]) {
            swap(arr, index, Math.round((index-1)/2));
            index = Math.round((index-1)/2);
        }
    }
    return arr;
}

function swap(arr, i, j) {
    let tmp = arr[i];
    arr[i] = arr[j];
    arr[j] = tmp; 
}

// 下沉操作
function heapify(arr, index, heapSize) {
    let left = index * 2 +1;

    while(left < heapSize) {
        let largest = (left+1<heapSize && arr[left+1] > arr[left])? left+1: left;

        largest = arr[largest]>arr[index]?largest:index;

        if (largest === index) {
            break;
        }

        swap(arr, index, largest);
        console.log('arr', arr)
        index = largest;
        left = 2*index+1;
    }
}


function heapSort(arr) {
    let heap = toBigHeap(arr);
    console.log('生成的堆', heap);
    let heapSize = heap.length;
    swap(heap, 0, --heapSize);
    while(heapSize> 0) {
        heapify(heap, 0,heapSize);
        swap(heap, 0, --heapSize);
    }
    
    return heap;
}

let arr = [1,3,4,5,6,7,2];

console.log(heapSort(arr))

File: code/src/main/java/com/xiong/learn/nowcoder/ch1/HeapSort.java
package com.xiong.learn.nowcoder.ch1;

public class HeapSort {

    public static void swap(int[] arr, int i, int j) {
        int t = arr[i];
        arr[i] = arr[j];
        arr[j] = t;
    }

    public static void heapInsert(int[] arr, int i) {
        while ((i - 1) / 2 >= 0 && arr[i] > arr[(i - 1) / 2]) {
            swap(arr, i, (i - 1) / 2);
            i = (i - 1) / 2;
        }
    }

    public static void heapify(int[] arr, int i, int size) {
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        while (l < size) {
            int largestIndex = r < size && arr[r] > arr[l] ? r : l;
            if (arr[i] < arr[largestIndex]) {
                swap(arr, i, largestIndex);
                i = largestIndex;
                l = 2 * i + 1;
                r = 2 * i + 2;
            } else {
                break;
            }
        }
    }

    public static void print(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            System.out.println(arr[i]);
        }
    }


    public static void heapSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            heapInsert(arr, i);
        }
        int size = arr.length;
        while (size > 0) {
            swap(arr, 0, size - 1);
            heapify(arr, 0, --size);
        }
    }

    public static void main(String[] args) {
        int[] arr = {1, 6, 5};
        heapSort(arr);
        for (int i = 0; i < arr.length; i++) {
            System.out.println(arr[i]);
        }
    }
}
